File: scripts/generate_sample_data.py
from pathlib import Path
import random
import string

import pyarrow as pa
import pyarrow.parquet as pq


def clean_old_parquet_files(output_dir: Path) -> None:
    """
    Deletes only old Parquet files from the output directory.
    """

    output_dir.mkdir(parents=True, exist_ok=True)

    for parquet_file in output_dir.glob("*.parquet"):
        parquet_file.unlink()


def random_text(length: int = 120) -> str:
    """
    Creates partly random text.
    Random text makes the input larger.
    Repeated patterns still allow compression in the output.
    """

    letters = string.ascii_letters + string.digits
    return "".join(random.choices(letters, k=length))


def generate_sample_parquet_files(
    output_dir: Path,
    number_of_files: int = 100,
    rows_per_file: int = 5000,
) -> None:
    """
    Generates demo Parquet data.

    Goal:
    - Input size around a few hundred MB
    - Output size clearly smaller after compaction
    """

    clean_old_parquet_files(output_dir)

    cities = ["Atlanta", "Kennesaw", "Marietta", "Dallas", "Austin"]
    statuses = ["COMPLETED", "PENDING", "FAILED", "REFUNDED"]
    categories = ["Electronics", "Books", "Clothing", "Grocery", "Sports"]

    repeated_note = (
        "Parquet compaction demo transaction record. "
        "This text repeats enough to allow output compression. "
    ) * 3

    total_size_bytes = 0

    for file_number in range(1, number_of_files + 1):
        transaction_ids = list(
            range(
                (file_number - 1) * rows_per_file,
                file_number * rows_per_file,
            )
        )

        customer_ids = [
            f"CUST_{random.randint(1, 2000)}"
            for _ in range(rows_per_file)
        ]

        city_values = [
            random.choice(cities)
            for _ in range(rows_per_file)
        ]

        status_values = [
            random.choice(statuses)
            for _ in range(rows_per_file)
        ]

        category_values = [
            random.choice(categories)
            for _ in range(rows_per_file)
        ]

        amounts = [
            round(random.uniform(10.0, 5000.0), 2)
            for _ in range(rows_per_file)
        ]

        # Mixed text:
        # repeated part helps compression
        # random part keeps input size realistic
        transaction_notes = [
            repeated_note + " REF_" + random_text(80)
            for _ in range(rows_per_file)
        ]

        table = pa.table(
            {
                "transaction_id": transaction_ids,
                "customer_id": customer_ids,
                "city": city_values,
                "status": status_values,
                "product_category": category_values,
                "amount": amounts,
                "transaction_notes": transaction_notes,
            }
        )

        output_file = output_dir / f"small_file_{file_number:03d}.parquet"

        # IMPORTANT:
        # Input files are intentionally uncompressed and inefficient.
        # This helps show size reduction after compaction.
        pq.write_table(
            table,
            output_file,
            compression=None,
            use_dictionary=False,
        )

        file_size_mb = output_file.stat().st_size / (1024 * 1024)
        total_size_bytes += output_file.stat().st_size

        print(f"Created: {output_file} | Size: {file_size_mb:.2f} MB")

    total_size_mb = total_size_bytes / (1024 * 1024)

    print()
    print("=" * 60)
    print(f"Generated {number_of_files} Parquet files")
    print(f"Rows per file: {rows_per_file}")
    print(f"Total rows: {number_of_files * rows_per_file}")
    print(f"Total input size: {total_size_mb:.2f} MB")
    print("=" * 60)


if __name__ == "__main__":
    generate_sample_parquet_files(
        output_dir=Path("sample_data/small_files"),
        number_of_files=100,
        rows_per_file=5000,
    )
    

File: src/bin/generate_sample_data.rs
use anyhow::Result;
use arrow_array::{ArrayRef, Float64Array, Int64Array, RecordBatch, StringArray};
use arrow_schema::{DataType, Field, Schema};
use parquet::arrow::arrow_writer::ArrowWriter;
use parquet::basic::Compression;
use parquet::file::properties::WriterProperties;
use rand::distributions::Alphanumeric;
use rand::{thread_rng, Rng};
use std::fs::{self, File};
use std::path::{Path, PathBuf};
use std::sync::Arc;

fn clean_old_parquet_files(output_dir: &Path) -> Result<()> {
    fs::create_dir_all(output_dir)?;

    for entry in fs::read_dir(output_dir)? {
        let entry = entry?;
        let path = entry.path();

        let is_parquet = path
            .extension()
            .and_then(|extension| extension.to_str())
            .map(|extension| extension.eq_ignore_ascii_case("parquet"))
            .unwrap_or(false);

        if is_parquet {
            fs::remove_file(path)?;
        }
    }

    Ok(())
}

fn random_text(length: usize) -> String {
    thread_rng()
        .sample_iter(&Alphanumeric)
        .take(length)
        .map(char::from)
        .collect()
}

fn build_schema() -> Arc<Schema> {
    Arc::new(Schema::new(vec![
        Field::new("transaction_id", DataType::Int64, false),
        Field::new("customer_id", DataType::Utf8, false),
        Field::new("city", DataType::Utf8, false),
        Field::new("status", DataType::Utf8, false),
        Field::new("product_category", DataType::Utf8, false),
        Field::new("amount", DataType::Float64, false),
        Field::new("transaction_notes", DataType::Utf8, false),
    ]))
}

fn generate_sample_parquet_files(
    output_dir: &Path,
    number_of_files: usize,
    rows_per_file: usize,
) -> Result<()> {
    clean_old_parquet_files(output_dir)?;

    let schema = build_schema();

    let cities = ["Atlanta", "Kennesaw", "Marietta", "Dallas", "Austin"];
    let statuses = ["COMPLETED", "PENDING", "FAILED", "REFUNDED"];
    let categories = ["Electronics", "Books", "Clothing", "Grocery", "Sports"];

    let repeated_note = "Parquet compaction demo transaction record. \
                         This repeated text helps demonstrate compression. ";

    let input_properties = WriterProperties::builder()
        .set_compression(Compression::UNCOMPRESSED)
        .set_dictionary_enabled(false)
        .build();

    let mut total_size_bytes = 0_u64;
    let mut rng = thread_rng();

    for file_number in 1..=number_of_files {
        let start_id = ((file_number - 1) * rows_per_file) as i64;
        let end_id = start_id + rows_per_file as i64;

        let transaction_ids = Int64Array::from_iter_values(start_id..end_id);

        let mut customer_ids = Vec::with_capacity(rows_per_file);
        let mut city_values = Vec::with_capacity(rows_per_file);
        let mut status_values = Vec::with_capacity(rows_per_file);
        let mut category_values = Vec::with_capacity(rows_per_file);
        let mut amounts = Vec::with_capacity(rows_per_file);
        let mut notes = Vec::with_capacity(rows_per_file);

        for _ in 0..rows_per_file {
            customer_ids.push(format!("CUST_{}", rng.gen_range(1..=2000)));

            city_values.push(cities[rng.gen_range(0..cities.len())].to_string());

            status_values.push(statuses[rng.gen_range(0..statuses.len())].to_string());

            category_values.push(categories[rng.gen_range(0..categories.len())].to_string());

            amounts.push(rng.gen_range(10.0..5000.0));

            notes.push(format!(
                "{} REF_{}",
                repeated_note.repeat(3),
                random_text(80)
            ));
        }

        let batch = RecordBatch::try_new(
            schema.clone(),
            vec![
                Arc::new(transaction_ids) as ArrayRef,
                Arc::new(StringArray::from(customer_ids)) as ArrayRef,
                Arc::new(StringArray::from(city_values)) as ArrayRef,
                Arc::new(StringArray::from(status_values)) as ArrayRef,
                Arc::new(StringArray::from(category_values)) as ArrayRef,
                Arc::new(Float64Array::from(amounts)) as ArrayRef,
                Arc::new(StringArray::from(notes)) as ArrayRef,
            ],
        )?;

        let output_file = output_dir.join(format!("small_file_{:03}.parquet", file_number));

        let file = File::create(&output_file)?;

        let mut writer = ArrowWriter::try_new(file, schema.clone(), Some(input_properties.clone()))?;

        writer.write(&batch)?;
        writer.close()?;

        let file_size = fs::metadata(&output_file)?.len();
        total_size_bytes += file_size;

        println!(
            "Created: {} | Size: {:.2} MB",
            output_file.display(),
            file_size as f64 / (1024_f64 * 1024_f64)
        );
    }

    println!();
    println!("============================================================");
    println!("Generated {} Parquet files", number_of_files);
    println!("Rows per file: {}", rows_per_file);
    println!("Total rows: {}", number_of_files * rows_per_file);
    println!(
        "Total input size: {:.2} MB",
        total_size_bytes as f64 / (1024_f64 * 1024_f64)
    );
    println!("============================================================");

    Ok(())
}

fn main() -> Result<()> {
    let output_dir = PathBuf::from("sample_data/small_files");

    generate_sample_parquet_files(&output_dir, 100, 5000)?;

    Ok(())
}